### DISCUSSION TITLE
referring to banks from C code

### DISCUSSION THREAD

**Discussion by mrdudz** [cc65/cc65#1499] [Ideas]
First of all, i know this is a solved problem on the assembler level. All of this can be done with external assembler files. However, it would still be nice if there would be a way to do it in plain C code - i admit i dont have a good idea how this could happen though.

So, the usecase is that i have "plugin" functions that are called via a table of function pointers. The various functions live in different banks and are called via a custom wrapper that then uses a table with the function pointers and banks in it. This wrapper is written in C because reasons :) Right now the table consists of function pointers and symbols that refer to their bank - and those symbols need to be maintained manually, which i'd like to avoid. The only idea i came up with so far would be writing some script that generates that table... but that is also something i'd like to avoid :)

So, perhaps someone has an idea how to make this happen in C code - some kind of extra "thing" must be introduced to get the  bank into the table.

Perhaps there is also a way to make regular function pointers work with wrapped-call - that seems really tricky if not plain impossible though, unfortunately.

Please refrain from telling "assembler is the answer" - i already know this. This is mostly an invite to think out of the box and then we *maybe* find a way to make it happen in C anyway :)

**Comment by groessler** [comment]
Maybe we could introduce some kind of "far pointer". Which is > 16 bits, and the upper byte refers to the bank.

If you grew up with 16-bit DOS programming, the concept should be intuitive :-)

**Reply by mrdudz** [reply]
yeah a addr:segment type of thing ... no idea how realistic this is to integrate into the compiler. the assembler also already has .near and .far - but no idea how this is related (65816 stuff i think)

**Comment by oliverschmidt** [comment]
> If you grew up with 16-bit DOS programming, the concept should be intuitive :-)

I liked `__based` pointers (https://docs.microsoft.com/en-us/cpp/cpp/based-pointers-cpp?view=msvc-160) and considered it pretty boring to have a flat 32-bit memory model when moving from Win16 to Win32 ;-)
